mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,909 @@
+// i.e. for a given constraint t(x) >= 0 returns -t(x) -1 >= 0
+SmallVector<int64_t, 8> complement(ArrayRef<int64_t> t);
+
+// shifts t by amount
+// i.e. for a given constraint t(x) >= 0 return t(x) + amount >= 0
+void shift(SmallVectorImpl<int64_t> &t, int amount);
+
+// add eq as two inequalities to target
+void addAsIneq(const ArrayRef<ArrayRef<int64_t>> eq,
+               SmallVectorImpl<ArrayRef<int64_t>> &target);
+
+// adds all Equalities to bs
+void addEqualities(FlatAffineConstraints &bs,
+                   const SmallVector<ArrayRef<int64_t>, 8> &equalities);
+
+// adds all Inequalities to bs
+void addInequalities(FlatAffineConstraints &bs,
+                     const SmallVector<ArrayRef<int64_t>, 8> &inequalities);
+
+// only gets called by classify
+// classifies all constraints into redundant, cut or adj_ineq according to the
+// ineqType that the simplex returns
+//
+// returns true if it has neither encountered a separate constraint nor more
+// than one adjacent inequality
+bool classify_ineq(Simplex &simp,
+                   const SmallVector<ArrayRef<int64_t>, 8> &constraints,
+                   Info &info);
+
+// classifies all constraints into redundant, cut, adj_ineq or t, where t stands
+// for a constraint adjacent to a the other polytope
+//
+// returns true if it has not encountered a separate constraint or more than one
+// adjcacent constraint
+bool classify(Simplex &simp,
+              const SmallVector<ArrayRef<int64_t>, 8> &inequalities,
+              const SmallVector<ArrayRef<int64_t>, 8> &equalities, Info &info);
+
+// compute the protrusionCase and return whether it has worked
+//
+// In the protusionCase, only CUT constraints and REDUNDANT constraints can
+// exist. This case differs from the CutCase in that it can still coalesce
+// polytopes, as long as they are not sticking out of each other by too much
+// (i.e. by less than 2). As we are considering an integer setting, the convex
+// hull of the union of those two polytopes actually is the same thing as the
+// union of the two polytope.
+//      _____          _____
+//    _|___  |        /     |
+//   | |   | |   ==> |      |
+//   | |   | |       |      |
+//   |_|___|_|       |______|
+//
+// TODO: find better example
+bool protrusionCase(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    Info &info_a, const Info &info_b, unsigned i, unsigned j);
+
+// compute, whether a constraint of cut sticks out of bs by more than 2
+bool stickingOut(const SmallVector<ArrayRef<int64_t>, 8> &cut,
+                 const FlatAffineConstraints &bs);
+
+// adds a FlatAffineConstraints and removes the sets at i and j.
+void addCoalescedBasicSet(
+    SmallVectorImpl<FlatAffineConstraints> &basicSetVector, unsigned i,
+    unsigned j, const FlatAffineConstraints &bs);
+
+// compute the cut case and return whether it has worked.
+//
+// The cut case is the case, for which a polytope only has REDUNDANT and CUT
```

Comment:
As far as I understand, the sloped cut constraint in the left diagram is not totally contained in the other polytope? What am I missing? Please elaborate in the documentation.

Reply:
The sloped cut constraint actually is contained within the left polytope. Note that due to them overlapping, I coudn't fully draw the left polytope. I think this is a limitation of the kind of representation and I will try to come up with a better example that is not limited like that.

Reply:
I changed the example, hopefully, it is clearer now. I tried to do the same example as the integer set coalescing paper uses for the cut case.